presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -383,26 +391,38 @@ private synchronized Optional<QueryResults> getCachedResult(long token)
     private synchronized QueryResults getNextResultWithRetry(long token, UriInfo uriInfo, String scheme, DataSize targetResultSize, boolean binaryResults)
     {
         QueryResults queryResults = getNextResult(token, uriInfo, scheme, targetResultSize, binaryResults);
-        if (queryResults.getError() == null || !queryResults.getError().isRetriable()) {
-            return queryResults;
-        }
 
-        // check if we have exceeded the global limit
-        retryCircuitBreaker.incrementFailure();
-        if (!retryCircuitBreaker.isRetryAllowed() || hasProducedResult) {
-            return queryResults;
+        boolean historyBasedOptimizationEnabled = useHistoryBasedPlanStatisticsEnabled(session) && trackHistoryBasedPlanStatisticsEnabled(session);
+        boolean hasNotRetried = queryManager.getQueryRetryCount(queryId) < 1;
+
+        if (historyBasedOptimizationEnabled && hasNotRetried && retryConditionsMet(queryResults) && retryQueryWithHistoryBasedOptimizationEnabled(session)) {
+            if (!originalBeforeRetryQueryId.isPresent() && !previousQueryTopLevelPlanHash.isPresent()) {
+                originalBeforeRetryQueryId = Optional.of(queryId);
+                previousQueryTopLevelPlanHash = getCurrentTopLevelPlanHash();
+                previousQueryFailureError = Optional.of(queryResults.getError());
+            }
         }
+        else if (queryManager.getQueryRetryCount(queryId) == 1 && retryQueryWithHistoryBasedOptimizationEnabled(session) && retryConditionsMet(queryResults)) {
+            if (planHashesEqual(getCurrentTopLevelPlanHash())) {
+                queryManager.failQuery(queryId, new PrestoException(GENERIC_INTERNAL_ERROR, "Since the plan hashes did not change, your retry query will not execute." +
```

Comment:
Ideally we should keep the retry transparent to the user. Is it possible to have the user only see the failure of the original query without awareness of the retry?

Reply:
@feilong-liu When the query is running they will see both queryIds as the retry takes it course. We can tweak the fail message, but the user will always see the retry in this case